Support batch size > 1 image-text inference

hiyouga opened the pull request:
# What does this PR do?

This PR follows #35558 https://github.com/huggingface/transformers/issues/40263

Consider a batch of image lists, where the first example has 1 image and the second example has 0 image. e.g.,

```python
images = [
  [Image],
  []
]
```

Using the latest code, it will receive a value error `Invalid input type. Must be a single image, a list of images, or a list of batches of images.`.

In this PR, we use `any` instead of `all` to judge if it is a valid nested list of images. Note that this behavior is the same as the one in transformers 4.48.0.

https://github.com/huggingface/transformers/blob/v4.48.0/src/transformers/models/mllama/image_processing_mllama.py#L535-L541

https://github.com/huggingface/transformers/blob/6bc0fbcfa7acb6ac4937e7456a76c2f7975fefec/src/transformers/models/mllama/image_processing_mllama.py#L535-L541

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@zucchini-nlp

hiyouga committed:
update make nested image list

hiyouga committed:
fix make flat list of images

hiyouga committed:
update type anno

hiyouga commented:
@zucchini-nlp Assuming the batch size is 2, we expect that the length of image lists should be the same is the batch size

zucchini-nlp reviewed (approved):
I see, makes sense. Also cc @yonigozlan since you added these functions, do you see any edge cases if we check `any`?

Otherwise LGTM

yonigozlan commented:
Hi @hiyouga ! Thanks for flagging this issue. I agree we should support inputs such as `[[image1], []]` Right now it seems to be causing some issues with SmolVLM processor, but this is more of a problem with SmolVLM than with this PR.

The issue I see is that we wouldn't catch an error now if we have  `[[image1], image2]` for example, when we should. But we cannot catch every possible wrong input formats, so this might not be too bad. WDYT @zucchini-nlp ?

zucchini-nlp commented:
@yonigozlan agreed, I think we can expect users to use consistent format within one input.

@hiyouga there's a failing test which is caused by this PR i think, can you take a look?

hiyouga committed:
fix image_processing_smolvlm

hiyouga commented:
Hi @zucchini-nlp , I have made necessary changes to `Gemma3ImageProcessor`, `Idefics2ImageProcessor`, `Idefics3ImageProcessor` and `SmolVLMImageProcessor`, to make them support inputs like `[[image], []]`

zucchini-nlp commented on tests/models/idefics2/test_processor_idefics2.py:
```diff
@@ -281,17 +281,11 @@ def test_process_interleaved_images_prompts_image_error(self):
         images = [self.image1, self.image2]
         with self.assertRaises(ValueError):
             processor(text=text, images=images, padding=True)
-        images = [self.image1]
-        with self.assertRaises(ValueError):
-            processor(text=text, images=images, padding=True)
```
didn't get why this doesn't throw error anymore, IMO passing flat images is ambiguous, and we throw errors instead of trying to infer which text corresponds to which image

zucchini-nlp reviewed (commented):
@hiyouga great, thanks for handling the tests! 

I see why we need to flatten images with the new changes, but i don't like calling it every time when one image is needed. I'd suggest to save one image in a variable at the beginning and add a small comment we we do that, so future us don't delete it  :)

hiyouga committed:
use first image

hiyouga commented:
@zucchini-nlp I have added a variable named `first_image_in_list` to indicate the one image we needed for getting input format.

> didn't get why this doesn't throw error anymore, IMO passing flat images is ambiguous, and we throw errors instead of trying to infer which text corresponds to which image

There are some logics in Idefics's processor that match the image with the text according to image tokens, so I decide to remove such cases in unittest.

https://github.com/huggingface/transformers/blob/v4.49.0/src/transformers/models/idefics3/processing_idefics3.py#L257-L268
https://github.com/huggingface/transformers/blob/a22a4378d97d06b7a1d9abad6e0086d30fdea199/src/transformers/models/idefics3/processing_idefics3.py#L257-L268

zucchini-nlp commented on src/transformers/models/idefics3/image_processing_idefics3.py:
```diff
@@ -713,6 +715,8 @@ def preprocess(
 
         # All transformations expect numpy arrays.
         images_list = [[to_numpy_array(image) for image in images] for images in images_list]
+        # Search for the first image in the image list.
+        first_image_in_list = make_flat_list_of_images(images_list)[0]
```
great, something more verbose would be better imo. Smth like

`Search for the first image in the image list. NOTE: we can't slice the fist image with `images_list[0][0]` in case if the first batch contains no images. See #36682 `

zucchini-nlp reviewed (approved):
Approving, a tiny nit about comments, overall LGTM. Thanks for handling it!

I will request one more review, then we can merge. Core code is not modified so we can ask for vision maintainer's review :)

hiyouga committed:
add verbose comment

hiyouga commented:
@zucchini-nlp Thanks for your review! Have adjusted the comments about this change

qubvel commented on src/transformers/models/gemma3/image_processing_gemma3.py:
```diff
@@ -357,16 +358,19 @@ def preprocess(
         images_list = [[to_numpy_array(image) for image in images] for images in images_list]
+        # Search for the first image in the image list.
+        # NOTE: we can't slice the first image with images_list[0][0] if the first batch contains no images. See #36682
+        first_image_in_list = make_flat_list_of_images(images_list)[0]
```
Looks a bit like overkill using this function with lots of validations under-the-hood to get the first element. Maybe try this one-liner. What do you think?

```
first_image_in_list = [images for images in images_list if images][0][0]
```

qubvel commented on src/transformers/image_utils.py:
```diff
@@ -243,7 +243,7 @@ def make_flat_list_of_images(
         isinstance(images, (list, tuple))
         and all(isinstance(images_i, (list, tuple)) for images_i in images)
-        and all(is_valid_list_of_images(images_i) for images_i in images)
+        and any(is_valid_list_of_images(images_i) for images_i in images)
```
hmm, I would better extend it to support an empty list, otherwise, we might have something not relevant in any of the lists, e.g.

```suggestion
        and all(is_valid_list_of_images(images_i) or not images_i  for images_i in images)
```

qubvel commented on src/transformers/image_utils.py:
```diff
@@ -277,7 +277,7 @@ def make_nested_list_of_images(
         isinstance(images, (list, tuple))
         and all(isinstance(images_i, (list, tuple)) for images_i in images)
-        and all(is_valid_list_of_images(images_i) for images_i in images)
+        and any(is_valid_list_of_images(images_i) for images_i in images)
```
same here

qubvel reviewed (commented):
Thanks for working on this @hiyouga, here are a few more comments 🤗

hiyouga committed:
fix images

hiyouga commented:
Hi @qubvel , thank you for review. I have updated the PR according to your suggestions. I also modified a key of the `pad` method because the previous one is ambiguous `images -> images_list`

qubvel commented:
We should keep `images` for backward compatibility and for consistency with other `pad` methods. However, we still can resolve the ambiguity with proper type hints and docstrings

hiyouga committed:
rollback

hiyouga commented:
@qubvel got it, how about the latest one?

hiyouga committed:
fix ut

hiyouga committed:
Merge branch 'main' into patch-14

hiyouga commented:
@qubvel Hi Pavel, I'd like to ask that is there anything needs to be resolved in this PR?

qubvel commented on src/transformers/models/smolvlm/image_processing_smolvlm.py:
```diff
@@ -561,11 +561,12 @@ def pad(
             else input_data_format
         )
         data_format = input_data_format if data_format is None else data_format
+        first_image_in_list = [images_ for images_ in images if images_][0][0]
```
Just nit, LGTM!

```suggestion
        # filter out empty image lists, then take first image of the first sample
        first_image_in_list = [sample_images for sample_images in images if sample_images][0][0]
```

hiyouga committed:
Update image_processing_smolvlm.py

hiyouga committed:
Update image_processing_idefics3.py

qubvel committed:
Merge branch 'main' into patch-14

hiyouga committed:
Merge branch 'main' into patch-14

qubvel committed:
Merge branch 'main' into patch-14

qubvel commented:
Waiting for the CI to be green to merge 😄

hiyouga commented:
@qubvel It seems that the integration of llama4 breaks all the processor unit tests 
![image](https://github.com/user-attachments/assets/79c1fceb-5e03-459b-82d4-b57d105dbd5e)

ArthurZucker reviewed (approved):
Can you documment what this enables? Like in the pipeline md?

hiyouga commented:
@ArthurZucker This PR mainly enables the `ImageTextToTextPipeline` to have both image-text and text-only inputs in a whole batch. However, I'm not sure where I should add the document. Could you provide some instructions?

ArthurZucker reviewed (commented):
sorry you are right its kind of obvious that it should support batch > 1 image, what I mean is to have a small doc example somewhere for people to play with it! Let's fix the conflicts and get this merged 🔥

zucchini-nlp committed:
merge main

zucchini-nlp committed:
add tests and fix some processors

zucchini-nlp committed:
fix copies

zucchini-nlp committed:
merge main

zucchini-nlp committed:
fix after rebase

zucchini-nlp committed:
make the test cover chat templates

zucchini-nlp committed:
sjip udop, no point in fixing it

zucchini-nlp commented:
@ArthurZucker let's merge this to fix VLM training in TRL

zucchini-nlp committed:
Merge branch 'main' into patch-14

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_36682). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

zucchini-nlp committed:
Merge branch 'main' into patch-14

zucchini-nlp committed:
fix after rebase

zucchini-nlp committed:
fix a few more tests

zucchini-nlp committed:
Merge branch 'main' into patch-14
